### PR TITLE
Add auto-resolution capability

### PR DESCRIPTION
**Pull request by pmjones**
This allows type hinted constructor params, that do not have explicit DI values, to receive new instances of their type hinted classes. It also allows explicit auto-resolution of type hints to a specific class or service. Incidentally, params type hinted to `array` will receive empty arrays when no explicit value is passed.

### CODE REVIEW

File: src/Container.php [hunk 1]
```diff
@@ -66,6 +66,15 @@ class Container implements ContainerInterface
 
     /**
      *
+     * A reference to the Factory $resolve.
+     *
+     * @var array
+     *
+     */
+    protected $resolve;
```
Comment (cordoval) [inline src/Container.php]: should be resolver? resolution? kind of confuses

Comment (pmjones) [inline src/Container.php]: /me nods

Yeah, totally open to other names here. I thought about various combinations of "auto" and "typehint" and "resolve" as well, but nothing really seems to capture the idea.

Comment (harikt) [inline src/Container.php]: Does changing to resolver means it will change to $di->resolver ?

Then I feel resolve itself is good.

Comment (pmjones) [inline src/Container.php]: Yes, it would mean $di->resolver .

**Comment by pmjones** [comment]
Ping @hidenorigoto @madapaja @djmattyg007 @tailwindsllc @szkshi @atakig @suzuyan @exsuzukishuhei @mnapoli @pies @bytephp @timw4mail @mosiyash @jleider @nubs @harikt @phpingguo @koriym @jblotus @isimmons @happyaccidents @yagmikita @jeremeamia @mihaeu @atelierdisko @kumamidori @Flugund @williamsodell @njasm @rjmackay @ushahidi

**Comment by pmjones** [comment]
Also ping @mbrevda and @adriengibrat who have explicitly asked for this.

**Comment by pmjones** [comment]
See also mailing list discussion at https://groups.google.com/forum/#!topic/auraphp/aeaEV-5Q76A

**Comment by mbrevda** [comment]
Thanks for remembering! Let me see if I can fit this in to the workflow somewhere...

**Comment by harikt** [comment]
I would like to ping @TomBZombie for he is doing bench marking of di containers .

**Comment by adriengibrat** [comment]
I'll be happy to update my benchmark :
https://github.com/adriengibrat/benchmarking-dependency-injection-containers
6 month ago the results was:
http://adriengibrat.github.io/benchmarking-dependency-injection-containers/

**Comment by djmattyg007** [comment]
I haven't yet had a chance to read over the patch (in fact, I haven't been keeping up with it for a few months now, so I've got a bit of catchup work to do), but from the description and the mailing list discussion, I like it quite a bit.

I do have one question: how is the callable type hint supported? "Not at all" is a perfectly good answer, I just wanted to make sure it'd been thought of and won't crash the code.

**Comment by pmjones** [comment]
@djmattyg007 As far as "callable" there's no way I can think of to auto-resolve it. There's no class associated with it, so we can't instantiate anything; and unlike "array" there's no one canonical callable representation. So I don't see that it _can_ be supported by auto-resolution. In short, "not at all" AFAICT. ;-)

File: src/Container.php (hunk 1 above)
Comment (pmjones) [inline src/Container.php]: Maybe `$di->typehint['ClassOrInterfaceName'] = $di->lazyGet('service_name')` would do the trick instead of `$di->resolve...`.

**Comment by harikt** [comment]
I am good with this. May be we press the green button :+1:  .

**Comment by pmjones** [comment]
Green button it is.

**Comment by cordoval** [comment]
:+1:

**Comment by mbrevda** [comment]
Were the docs updated?

**Comment by pmjones** [comment]
No, that's the very next task. I think a full README rewrite is in order at this point.

**Comment by pmjones** [comment]
FWIW all the pre-existing idioms work unchanged.

**Comment by mbrevda** [comment]
I didnt say "requires" :). And yes, the current readme is a bit of a
ramble...

**Comment by adriengibrat** [comment]
Benchmark updated:
http://adriengibrat.github.io/benchmarking-dependency-injection-containers/

**Comment by pmjones** [comment]
That's very heartening. :-)

**Comment by harikt** [comment]
:+1:

**Pull request merged**